File: app/src/main/java/com/example/android/simpleweather/WeatherRepository.java
package com.example.android.simpleweather;

import android.util.Log;

import com.example.android.simpleweather.dto.CurrentForecast;
import com.example.android.simpleweather.dto.WeatherForecast;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

class WeatherRepository {

    @Inject
    public WeatherRepository() {}

    private final OpenWeatherDataSource service = RetrofitClientInstance.getRetrofit().create(OpenWeatherDataSource.class);

    public void getWeatherData(double latitude, double longitude, Callback<WeatherForecast> responseCallback) {
        Call<WeatherForecast> call = service.getWeather(latitude, longitude, BuildConfig.OPEN_WEATHER_KEY);
        call.enqueue(responseCallback);

        // TODO: see if exception is possible here, and deal with it if needed
    }

    public void getCurrentWeatherData(double latitude, double longitude, Callback<CurrentForecast> responseCallback) {
        Call<CurrentForecast> call = service.getCurrentWeather(latitude, longitude, BuildConfig.OPEN_WEATHER_KEY);
        call.enqueue(responseCallback);

        // TODO: see if exception is possible here, and deal with it if needed
    }
}

File: app/src/main/java/com/example/android/simpleweather/WeatherRepository.kt
package com.example.android.simpleweather

import android.util.Log
import com.example.android.simpleweather.dto.WeatherForecast
import java.lang.IllegalStateException
import javax.inject.Inject

class WeatherRepository @Inject constructor() {
    private val service: OpenWeatherDataSource = RetrofitClientInstance.retrofit.create(OpenWeatherDataSource::class.java)

    suspend fun getWeatherData(latitude: Double, longitude: Double): WeatherForecast {
        try {
            val response = service.listRepos(latitude, longitude, BuildConfig.OPEN_WEATHER_KEY)
            val body = response.body()
            if (body != null) {
                return body
            }
        } catch (e: Exception) {
            Log.e("WeatherRepository", e.message.toString())
            throw IllegalStateException(e)
        }
        throw IllegalStateException()
        // https://api.openweathermap.org/data/2.5/onecall?lat=42.1&lon=-73.1&appid=681089bf5b3e811ed4584a8f55fc5120
    }
}
